File: src/cpj/ipc/SharedMemory.java
package cpj.ipc;

import java.io.RandomAccessFile;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.charset.StandardCharsets;
import com.fasterxml.jackson.databind.ObjectMapper;

public class SharedMemory implements AutoCloseable {
    private final RandomAccessFile file;
    private final MappedByteBuffer buffer;
    private final ObjectMapper jsonMapper;
    private final String name;
    private final long size;

    public SharedMemory(String name, long size, boolean create) throws Exception {
        this.name = name;
        this.size = size;
        this.jsonMapper = new ObjectMapper();

        // Open or create shared memory file
        file = new RandomAccessFile("/dev/shm/" + name, "rw");
        if (create) {
            file.setLength(size);
        }

        // Map the file into memory
        buffer = file.getChannel().map(
                FileChannel.MapMode.READ_WRITE, 0, size);
    }

    public void write(byte[] data, int offset) {
        buffer.position(offset);
        buffer.put(data);
    }

    public byte[] read(int size, int offset) {
        buffer.position(offset);
        byte[] data = new byte[size];
        buffer.get(data);
        return data;
    }

    public void writeJson(Object data, int offset) throws Exception {
        byte[] jsonBytes = jsonMapper.writeValueAsBytes(data);
        // Write size prefix followed by JSON data
        buffer.position(offset);
        buffer.putInt(jsonBytes.length);
        buffer.put(jsonBytes);
    }

    public <T> T readJson(Class<T> valueType, int offset) throws Exception {
        buffer.position(offset);
        int size = buffer.getInt();
        byte[] jsonBytes = new byte[size];
        buffer.get(jsonBytes);
        return jsonMapper.readValue(jsonBytes, valueType);
    }

    @Override
    public void close() throws Exception {
        file.close();
    }

    public static void main(String[] args) {
        try (SharedMemory shm = new SharedMemory("test", 1024, true)) {
            // Test basic data
            byte[] testData = "Hello from Java!".getBytes(StandardCharsets.UTF_8);
            shm.write(testData, 0);
            byte[] readData = shm.read(testData.length, 0);
            System.out.println("Basic data test: " + new String(readData));

            // Test JSON data
            TestData testJson = new TestData();
            testJson.message = "Hello";
            testJson.numbers = new int[] { 1, 2, 3 };

            shm.writeJson(testJson, 0);
            TestData readJson = shm.readJson(TestData.class, 0);
            System.out.println("JSON data test: " + readJson);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    // Test class for JSON serialization
    public static class TestData {
        public String message;
        public int[] numbers;

        @Override
        public String toString() {
            return String.format("TestData{message='%s', numbers=%s}",
                    message, java.util.Arrays.toString(numbers));
        }
    }
}

File: src/ipc/shared_memory.py
"""
CPJ IPC Module - Python Bindings for Shared Memory
"""
import mmap
import os
import struct
from typing import Optional, Any
import json

class SharedMemory:
    def __init__(self, name: str, size: int, create: bool = False):
        self.name = name
        self.size = size
        
        if create:
            self.fd = os.open(f"/dev/shm/{name}", os.O_CREAT | os.O_RDWR, 0o666)
            os.truncate(self.fd, size)
        else:
            self.fd = os.open(f"/dev/shm/{name}", os.O_RDWR)
            
        self.mmap = mmap.mmap(self.fd, size)
    
    def __del__(self):
        if hasattr(self, 'mmap'):
            self.mmap.close()
        if hasattr(self, 'fd'):
            os.close(self.fd)
    
    def write(self, data: bytes, offset: int = 0) -> None:
        """Write bytes to shared memory at specified offset"""
        self.mmap[offset:offset + len(data)] = data
    
    def read(self, size: int, offset: int = 0) -> bytes:
        """Read bytes from shared memory at specified offset"""
        return bytes(self.mmap[offset:offset + size])
    
    def write_json(self, data: Any, offset: int = 0) -> None:
        """Write JSON-serializable data to shared memory"""
        json_bytes = json.dumps(data).encode('utf-8')
        # Write size prefix followed by data
        self.write(struct.pack('!I', len(json_bytes)), offset)
        self.write(json_bytes, offset + 4)
    
    def read_json(self, offset: int = 0) -> Any:
        """Read JSON data from shared memory"""
        size = struct.unpack('!I', self.read(4, offset))[0]
        json_bytes = self.read(size, offset + 4)
        return json.loads(json_bytes.decode('utf-8'))

# Example usage
if __name__ == '__main__':
    # Create shared memory
    shm = SharedMemory("test", 1024, create=True)
    
    # Write and read basic data
    test_data = b"Hello from Python!"
    shm.write(test_data)
    read_data = shm.read(len(test_data))
    print(f"Basic data test: {read_data.decode()}")
    
    # Write and read JSON data
    test_json = {
        "message": "Hello",
        "data": [1, 2, 3],
        "nested": {"key": "value"}
    }
    shm.write_json(test_json)
    read_json = shm.read_json()
    print(f"JSON data test: {read_json}")